src/NuGet.Clients/NuGet.Credentials/CredentialService.cs

```diff
@@ -158,21 +174,21 @@ public class CredentialService : ICredentialService
             return _providerCredentialCache.TryGetValue(key, out credentials);
         }
 
-        private void AddToCredentialCache(Uri uri, bool isProxy, ICredentialProvider provider,
+        private void AddToCredentialCache(Uri uri, CredentialRequestType type, ICredentialProvider provider,
             CredentialResponse credentials)
         {
-            _providerCredentialCache[CredentialCacheKey(uri, isProxy, provider)] = credentials;
+            _providerCredentialCache[CredentialCacheKey(uri, type, provider)] = credentials;
         }
 
-        private static string RetryCacheKey(Uri uri, bool isProxy, ICredentialProvider provider)
+        private static string RetryCacheKey(Uri uri, CredentialRequestType type, ICredentialProvider provider)
         {
-            return $"{provider.Id}_{isProxy}_{uri}";
+            return $"{provider.Id}_{type}_{uri}";
```

Comment:
Correct me if I'm wrong but I think cache key should be the same for 401 and for 403.